SmartTutor/ContentModel/LearningObjects/ChallengeModel/FulfillmentStrategy/MetricChecker/BasicMetricsChecker.cs

```diff
@@ -11,14 +12,18 @@ public class BasicMetricsChecker : ChallengeFulfillmentStrategy
         public List<MetricRangeRule> ClassMetricRules { get; set; }
         public List<MetricRangeRule> MethodMetricRules { get; set; }
 
-        public BasicMetricsChecker() {}
+        private List<MetricHint> _metricHints;
 
-        public BasicMetricsChecker(List<MetricRangeRule> classMetricRules, List<MetricRangeRule> methodMetricRules)
+        public BasicMetricsChecker() { }
+
+        public BasicMetricsChecker(List<MetricRangeRule> classMetricRules, List<MetricRangeRule> methodMetricRules, List<MetricHint> metricHints)
         {
             ClassMetricRules = classMetricRules;
             MethodMetricRules = methodMetricRules;
+            _metricHints = metricHints;
         }
 
+        //TODO: return object (with hints), not just bool
         public override bool CheckChallengeFulfillment(List<CaDETClass> solutionAttempt)
```

Comment:
I would return an object of type "ChallengeEvaluation" with the following fields:
- bool: ChallengeCompleted
- List<ChallengeHint> ApplicableHints
 (note the name - ChallengeHint)

When a challenge is complete, we can offer all hints if the trainee wants to examine them. Otherwise, we can offer hints that make sense. I've been thinking about who can offer these hints best, and it might be the low-level rules (in this case the MetricRule). For example, a MetricRule that counts the number of magic numbers might provide a hint for removing magic numbers. Basically, the MetricRule should return a Hint, and the BasicMetricChecker can aggregate these hints into a ChallengeEvaluation.
Does this make sense?

For the NameCheckerStrategy, we can define NamingRules that each contain a list of words to avoid and words to include. Each rule can be focused to return a Hint when it is not satisfied.

Reply:
It does. I have separated the two types of hints: applicable and metric based. Hints that are being generated are moved to MetricRule, and I will upgrade them so they can give appropriate assistance for trainees.